src/Toolbox.cpp

```diff
@@ -0,0 +1,412 @@
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details. 
+ * 
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+**/
+#include "Toolbox.h"
+#include "ZuluSCSI_disk.h"
+#include "ZuluSCSI_cdrom.h"
+#include "ZuluSCSI_log.h"
+#include <minIni.h>
+#include <SdFat.h>
+extern "C" {
+#include <toolbox.h>
+#include <scsi2sd_time.h>
+#include <sd.h>
+#include <mode.h>
+}
+
+
+extern "C" int8_t scsiToolboxEnabled()
+{
+    static int8_t enabled = -1;
+    if (enabled == -1)
+    {
+        enabled = ini_getbool("SCSI", "EnableToolbox", 0, CONFIGFILE);
+    }
+    return enabled == 1;
+}
+
+
+static bool toolboxFilenameValid(const char* name, bool isCD = false)
+{
+    if(strlen(name) == 0)
+    {
+        dbgmsg("toolbox: Ignoring filename empty file name");
+        return false;
+    }
+    if (isCD)
+    {
+      return scsiDiskFilenameValid(name);
+    }
+    return true;
+}
+
+static void doCountFiles(const char * dir_name, bool isCD = false)
+{
+    FsFile dir;
+    FsFile file;
+    char name[MAX_FILE_PATH] = {0};
+    dir.open(dir_name);
+    dir.rewindDirectory();
+    uint8_t file_count = 0;
+    while (file.openNext(&dir, O_RDONLY))
+    {
+        if(file.getError() > 0)
+        {
+            file.close();
+            break;
+        }
+        bool isDir = file.isDirectory();
+        file.getName(name, MAX_FILE_PATH);
+        file.close();
+        if (isCD && isDir)
+            continue;
+        // only count valid files.
+        if(toolboxFilenameValid(name, isCD))
+        {
+            file_count = file_count + 1;
+            if(file_count > 100) {
+                scsiDev.status = CHECK_CONDITION;
+                scsiDev.target->sense.code = ILLEGAL_REQUEST;
+                scsiDev.target->sense.asc = OPEN_RETRO_SCSI_TOO_MANY_FILES;
+                scsiDev.phase = STATUS;
+                dir.close();
+                return;
+            }
+        }
+  }
+  scsiDev.data[0] = file_count;
+  scsiDev.dataLen = sizeof(file_count);
+  scsiDev.phase = DATA_IN;
+}
+
+static void onListFiles(const char * dir_name, bool isCD = false) {
+    FsFile dir;
+    FsFile file;
+
+    memset(scsiDev.data, 0, 4096);
+    int ENTRY_SIZE = 40;
+    char name[MAX_FILE_PATH] = {0};
+    dir.open(dir_name);
+    dir.rewindDirectory();
+    uint8_t index = 0;
+    uint8_t file_entry[40] = {0};
+    while (file.openNext(&dir, O_RDONLY))
+    {   
+        uint8_t isDir = file.isDirectory() ? 0x00 : 0x01;
+        int len = file.getName(name, MAX_FILE_PATH);
+        if (len > MAX_MAC_PATH)
+            name[MAX_MAC_PATH] = 0x0;
+        uint64_t size = file.fileSize();
+        file.close();
+        if (!toolboxFilenameValid(name, isCD))
+            continue;
```

Comment:
This `toolboxFilenameValid` check in `onListFiles` is different from the one in `doCountFiles`. When listing files, the filename is truncated to MAX_MAC_PATH characters before checking for being value, but when counting files the filename is not truncated. So the count could include files that would be excluded in the listing, if their filenames are too long and the extension gets cut off.

Reply:
`get_file_from_index` also does not truncate the filenames, so it counts indexes different from `onListFiles` too.